common/hashing.h

```diff
@@ -0,0 +1,721 @@
+  // multiplication. However, it is not *necessary* as we do capture the
+  // complete 128-bit result. Where reasonable, we XOR random data into operands
+  // here to try and increase the distribution of bits feeding the multiply.
+  static auto Mix(uint64_t lhs, uint64_t rhs) -> uint64_t;
+
+  // We have a 64-byte random data pool designed to fit on a single cache line.
+  // This routine allows sampling it at byte indices, which allows getting 64 -
+  // 8 different random 64-bit results. The offset must be in the range [0, 56).
+  static auto SampleRandomData(ssize_t offset) -> uint64_t {
+    CARBON_DCHECK(offset + sizeof(uint64_t) < sizeof(StaticRandomData));
+    uint64_t data;
+    memcpy(&data,
+           reinterpret_cast<const unsigned char*>(&StaticRandomData) + offset,
+           sizeof(data));
+    return data;
+  }
+
+  // A throughput-optimized routine for when the byte sequence size is
+  // guaranteed to be >32.
+  static auto HashSizedBytesLarge(Hasher hash, llvm::ArrayRef<std::byte> bytes)
+      -> Hasher;
+
+  // Random data taken from the hexadecimal digits of Pi's fractional component,
+  // written in lexical order for convenience of reading. The resulting
+  // byte-stream will be different due to little-endian integers. These can be
+  // used directly for convenience rather than calling `SampleRandomData`, but
+  // be aware that this is the underlying pool. The goal is to reuse the same
+  // single cache-line of constant data.
+  //
+  // The initializers here can be generated with the following shell script,
+  // which will generate 8 64-bit values and one more digit. The `bc` command's
+  // decimal based scaling means that without getting at least some extra hex
+  // digits rendered there will be rounding that we don't want so the script
+  // below goes on to produce one more hex digit ensuring the the 8 initializers
+  // aren't rounded in any way. Using a higher scale won't cause the 8
+  // initializers here to change further.
+  //
+  // ```sh
+  // echo 'obase=16; scale=155; 4*a(1)' | env BC_LINE_LENGTH=500 bc -l \
+  //  | cut -c 3- | tr '[:upper:]' '[:lower:]' \
+  //  | sed -e "s/.\{4\}/&'/g" \
+  //  | sed -e "s/\(.\{4\}'.\{4\}'.\{4\}'.\{4\}\)'/0x\1,\n/g"
+  // ```
+  static inline constexpr std::array<uint64_t, 8> StaticRandomData = {
+      0x243f'6a88'85a3'08d3, 0x1319'8a2e'0370'7344, 0xa409'3822'299f'31d0,
+      0x082e'fa98'ec4e'6c89, 0x4528'21e6'38d0'1377, 0xbe54'66cf'34e9'0c6c,
+      0xc0ac'29b7'c97c'50dd, 0x3f84'd5b5'b547'0917,
+  };
+
+ private:
+  template <typename T>
+  friend auto HashValue(const T& value, uint64_t seed) -> HashCode;
+  template <typename T>
+  friend auto HashValue(const T& value) -> HashCode;
+
+  explicit Hasher(uint64_t seed) : buffer(seed) {}
+
+  Hasher() = default;
+
+  // The multiplicative hash constant from Knuth, derived from 2^64 / Phi.
+  static constexpr uint64_t MulConstant = 0x9e37'79b9'7f4a'7c15U;
+
+  uint64_t buffer;
+};
+
+namespace Detail {
```

Comment:
I assume the intent is that this is where we put the `CarbonHash` overloads that can't be found by ADL? If so, it might be a good idea to document that, so people don't put overloads for types in the `Carbon` namespace here.